File: module4/depthFirstValues.js
/*
depth first values

Write a function, depthFirstValues, that takes in the root of a binary tree. The function should return an array containing all values of the tree in depth-first order.

Hey. This is our first binary tree problem, so you should be liberal with watching the Approach and Walkthrough. Be productive, not stubborn. -AZ
test_00:

const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f

depthFirstValues(a);
//    -> ['a', 'b', 'd', 'e', 'c', 'f']

test_01:

const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');
const g = new Node('g');

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;
e.left = g;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f
//    /
//   g

depthFirstValues(a);
//    -> ['a', 'b', 'd', 'e', 'g', 'c', 'f']
*/

// class Node {
//   constructor(val) {
//     this.val = val;
//     this.left = null;
//     this.right = null;
//   }
// }

// my recursive approach
const depthFirstValues = (root, arr = []) => {
  if (root === null) return [];
  arr.push(root.val);
  if (root.left) depthFirstValues(root.left, arr);
  if (root.right) depthFirstValues(root.right, arr);
  return arr;
};

// my iterative approach
const depthFirstValues = (root) => {
  if (root === null) return [];
  let stack = [root];
  let result = [];

  while (stack.length) {
    candidate = stack.pop();
    result.push(candidate.val);

    if (candidate.right) stack.push(candidate.right);
    if (candidate.left) stack.push(candidate.left);
  }

  return result;
};

File: module4/depth_first_values.py
'''
depth first values

Write a function, depth_first_values, that takes in the root of a binary tree. The function should return a list containing all values of the tree in depth-first order.

Hey. This is our first binary tree problem, so you should be liberal with watching the Approach and Walkthrough. Be productive, not stubborn. -AZ
test_00:

a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')
a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#      a
#    /   \
#   b     c
#  / \     \
# d   e     f

depth_first_values(a)
#   -> ['a', 'b', 'd', 'e', 'c', 'f']

test_01:

a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')
g = Node('g')
a.left = b
a.right = c
b.left = d
b.right = e
c.right = f
e.left = g

#      a
#    /   \
#   b     c
#  / \     \
# d   e     f
#    /
#   g

depth_first_values(a)
#   -> ['a', 'b', 'd', 'e', 'g', 'c', 'f']
'''

# class Node:
#   def __init__(self, val):
#     self.val = val
#     self.left = None
#     self.right = None

# my approach
# iterative
def depth_first_values(root):
  if root is None:
    return []
  stack = [ root ]
  result = []

  while len(stack):
    current = stack.pop()
    result.append(current.val)

    if current.right:
      stack.append(current.right)
    if current.left:
      stack.append(current.left)
  return result

# recursive
def depth_first_values(root):
  if root is None:
    return []
  leftValues = depth_first_values(root.left)
  rightValues = depth_first_values(root.right)
  return [root.val, *leftValues, *rightValues]